### PR TITLE
Add MAX_PODS env allowing kubelet to be max-pods aware.

### PR DESCRIPTION
**Pull request by grayluck**
**What this PR does / why we need it**:
Add MAX_PODS env allowing kubelet to be max-pods aware. This env will be one of the argument of kubelet. By setting this parameter, correct number of pod capacity for the node will be reported.

**Special notes for your reviewer**:
/assign @MrHohn @jpbetz 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cluster/gce/util.sh [hunk 1]
```diff
@@ -1136,6 +1139,11 @@ EOF
   if [ -n "${SCHEDULING_ALGORITHM_PROVIDER:-}" ]; then
     cat >>$file <<EOF
 SCHEDULING_ALGORITHM_PROVIDER: $(yaml-quote ${SCHEDULING_ALGORITHM_PROVIDER})
+EOF
+  fi
+  if [ -n "${MAX_PODS:-}" ]; then
+    cat >>$file <<EOF
+MAX_PODS: $(yaml-quote ${MAX_PODS})
```
Comment (MrHohn) [inline cluster/gce/util.sh]: Should this env be more specific? e.g. `MAX_PODS_PER_NODE`, `MAX_PODS_PER_KUBELET`?

**Comment by MrHohn** [comment]
cc @satyasm

File: cluster/gce/util.sh (hunk 1 above)
Comment (grayluck) [reply cluster/gce/util.sh]: Renamed. MAX_PODS_PER_NODE picked since this env might be used somewhere else in the future, not only for kubelet.

**Review by MrHohn — APPROVED** [review]
/lgtm

**Comment by satyasm** [comment]
if we are introducing this variable, then I believe we should also update https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/config-default.sh#L293 to make sure that the cluster as a whole is consistent based on the given MAX_POD_PER_NODE value.

File: cluster/gce/config-default.sh [hunk 2]
```diff
@@ -305,6 +305,8 @@ if [ ${ENABLE_IP_ALIASES} = true ]; then
   PROVIDER_VARS="${PROVIDER_VARS:-} ENABLE_IP_ALIASES"
   PROVIDER_VARS="${PROVIDER_VARS:-} NODE_IPAM_MODE"
   PROVIDER_VARS="${PROVIDER_VARS:-} SECONDARY_RANGE_NAME"
+  # Number of Pods that can run on this node.
+  MAX_PODS_PER_NODE=${MAX_PODS_PER_NODE:-110}
```
Comment (MrHohn) [inline cluster/gce/config-default.sh]: Please add this to cluster/gce/config-test.sh as well.

**Comment by grayluck** [comment]
Sorry wrong assigns.
/unassign @jpbetz @MaciekPytel

File: cluster/gce/config-default.sh (hunk 2 above)
Comment (grayluck) [reply cluster/gce/config-default.sh]: Done. Squashed BTW.

**Comment by satyasm** [comment]
Also, we should only honor this flag when ENABLE_IP_ALIAS is set to true and error out otherwise. Let's not merge it in till we can co-ordinate both the --max-pods argument to kubelet and the alias range size based on the MAX_PODS_PER_NODE environment. Changing only one or the other could lead to a broken cluster configuration. Thanks!

**Comment by MrHohn** [comment]
From https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/63114/pull-kubernetes-e2e-gce/32675/.

Saw this error in log:
```
W0426 05:24:02.592] 2018/04/26 05:24:02 process.go:150: Running: ./hack/e2e-internal/e2e-down.sh
W0426 05:24:02.607] ./hack/e2e-internal/../../cluster/../cluster/gce/../../cluster/gce/config-test.sh: line 317: MAX_PODS_PER_NODE: unbound variable
```

File: cluster/gce/config-default.sh [hunk 3]
```diff
@@ -289,8 +289,10 @@ ENABLE_RESCHEDULER="${KUBE_ENABLE_RESCHEDULER:-true}"
 ENABLE_IP_ALIASES=${KUBE_GCE_ENABLE_IP_ALIASES:-false}
 NODE_IPAM_MODE=${KUBE_GCE_NODE_IPAM_MODE:-RangeAllocator}
 if [ ${ENABLE_IP_ALIASES} = true ]; then
-  # Size of ranges allocated to each node. Currently supports only /32 and /24.
```
Comment (MrHohn) [inline cluster/gce/config-default.sh]: > Currently supports only /32 and /24.

Sorry if this is known, has this limitation been lifted yet? I'm still seeing it on https://cloud.google.com/vpc/docs/alias-ip. @satyasm

Comment (MrHohn) [reply cluster/gce/config-default.sh]: Confirmed this is not an issue anymore. Thanks @satyasm for verifying!

File: cluster/gce/config-common.sh [hunk 4]
```diff
@@ -98,6 +98,14 @@ function get-cluster-ip-range {
   echo "${suggested_range}"
 }
 
+function get-alias-range-size() {
+  for pow in {0..8}; do
```
Comment (satyasm) [inline cluster/gce/config-common.sh]: Why just 0..8? We should go all the way to 31, as we are not limiting the max pod size.

File: cluster/gce/config-common.sh [hunk 5]
```diff
@@ -98,6 +98,14 @@ function get-cluster-ip-range {
...
+    if (( 1 << $pow > $1 * 2 )); then
+      echo $((32 - pow))
```
Comment (satyasm) [inline cluster/gce/config-common.sh]: why use echo as the return style? Isn't the normal mechanism to set a VARIABLE and then use that?

File: cluster/gce/config-default.sh [hunk 6]
```diff
@@ -289,8 +289,10 @@ ENABLE_RESCHEDULER="${KUBE_ENABLE_RESCHEDULER:-true}"
...
-  IP_ALIAS_SIZE=${KUBE_GCE_IP_ALIAS_SIZE:-/24}
+  # Number of Pods that can run on this node.
+  MAX_PODS_PER_NODE=${MAX_PODS_PER_NODE:-110}
+  # Size of ranges allocated to each node.
+  IP_ALIAS_SIZE="/$(get-alias-range-size MAX_PODS_PER_NODE)"
```
Comment (satyasm) [inline cluster/gce/config-default.sh]: see earlier comment about echo v/s using normal variables to get this computed value.

File: cluster/gce/config-test.sh [hunk 7]
```diff
@@ -296,8 +296,10 @@ ENABLE_RESCHEDULER="${KUBE_ENABLE_RESCHEDULER:-true}"
 ENABLE_IP_ALIASES=${KUBE_GCE_ENABLE_IP_ALIASES:-false}
 NODE_IPAM_MODE=${KUBE_GCE_NODE_IPAM_MODE:-RangeAllocator}
 if [ ${ENABLE_IP_ALIASES} = true ]; then
-  # Size of ranges allocated to each node. gcloud current supports only /32 and /24.
-  IP_ALIAS_SIZE=${KUBE_GCE_IP_ALIAS_SIZE:-/24}
+  # Number of Pods that can run on this node.
+  MAX_PODS_PER_NODE=${MAX_PODS_PER_NODE:-110}
+  # Size of ranges allocated to each node.
+  IP_ALIAS_SIZE="/$(get-alias-range-size MAX_PODS_PER_NODE)"
```
Comment (satyasm) [inline cluster/gce/config-test.sh]: see earlier comment on echo v/s using variables for return

File: cluster/gce/config-common.sh (hunk 4 above)
Comment (grayluck) [reply cluster/gce/config-common.sh]: You're right I shouldn't set this limit here.

File: cluster/gce/config-common.sh (hunk 5 above)
Comment (grayluck) [reply cluster/gce/config-common.sh]: I am just following the code style in the file cluster/gce/config-common.sh. Also returning with echo can eliminate unnecessary variable existance assumptions.

File: cluster/gce/config-common.sh [hunk 8]
```diff
@@ -98,6 +98,14 @@ function get-cluster-ip-range {
   echo "${suggested_range}"
 }
 
+function get-alias-range-size() {
```
Comment (bowei) [inline cluster/gce/config-common.sh]: Document this function, including  parameters

File: cluster/gce/config-default.sh (hunk 6 above)
Comment (bowei) [inline cluster/gce/config-default.sh]: does this work? don't you need ${MAX_PODS_PER_NODE} here?

**Comment by grayluck** [comment]
Manual test passed.

### Env:
MAX_PODS_PER_NODE=30
ENABLE_IP_ALIASES=true
KUBE_GCE_ENABLE_IP_ALIASES=true

### Command:
`go run hack/e2e.go -- --build --up`

### Result:

#### kubectl get node e2e-test-yankaiz-minion-group-qqgx -o yaml
```
spec:
  podCIDR: 10.64.0.128/26
  providerID: gce://yankaiz-kube-test/us-central1-b/e2e-test-yankaiz-minion-group-qqgx
status:
  allocatable:
    cpu: "2"
    ephemeral-storage: "91117161526"
    hugepages-2Mi: "0"
    memory: 7402196Ki
    pods: "30"
  capacity:
    cpu: "2"
    ephemeral-storage: 98868448Ki
    hugepages-2Mi: "0"
    memory: 7658196Ki
    pods: "30"
```

### gcloud compute ssh e2e-test-yankaiz-minion-group-qqgx
`yankaiz@e2e-test-yankaiz-minion-group-qqgx ~ $ cat /proc/1324/cmdline
/home/kubernetes/bin/kubelet--v=4--serialize-image-pulls=false--allow-privileged=true--cloud-provider=gce--experimental-mounter-path=/home/kubernetes/containerized_mounter/mounter--experimental-check-node-capabilities-before-mount=true--cert-dir=/var/lib/kubelet/pki/--experimental-kernel-memcg-notification=true--bootstrap-kubeconfig=/var/lib/kubelet/bootstrap-kubeconfig--kubeconfig=/var/lib/kubelet/kubeconfig--cni-bin-dir=/home/kubernetes/bin--network-plugin=kubenet--non-masquerade-cidr=0.0.0.0/0--volume-plugin-dir=/home/kubernetes/flexvolume--node-labels=beta.kubernetes.io/fluentd-ds-ready=true,beta.kubernetes.io/metadata-proxy-ready=true--container-runtime=docker--max-pods=30--config/home/kubernetes/kubelet-config.yaml`

Got --max-pods=30 as a flag given to kubelet.

File: cluster/gce/config-default.sh (hunk 6 above)
Comment (grayluck) [reply cluster/gce/config-default.sh]: It works in manual test.

File: cluster/gce/config-common.sh (hunk 8 above)
Comment (grayluck) [reply cluster/gce/config-common.sh]: Done.

File: cluster/gce/config-default.sh [hunk 9]
```diff
@@ -289,8 +289,10 @@ ENABLE_RESCHEDULER="${KUBE_ENABLE_RESCHEDULER:-true}"
 ENABLE_IP_ALIASES=${KUBE_GCE_ENABLE_IP_ALIASES:-false}
 NODE_IPAM_MODE=${KUBE_GCE_NODE_IPAM_MODE:-RangeAllocator}
 if [ ${ENABLE_IP_ALIASES} = true ]; then
-  # Size of ranges allocated to each node. Currently supports only /32 and /24.
-  IP_ALIAS_SIZE=${KUBE_GCE_IP_ALIAS_SIZE:-/24}
+  # Number of Pods that can run on this node.
+  MAX_PODS_PER_NODE=${MAX_PODS_PER_NODE:-110}
+  # Size of ranges allocated to each node.
+  IP_ALIAS_SIZE="/$(get-alias-range-size $MAX_PODS_PER_NODE)"
```
Comment (bowei) [inline cluster/gce/config-default.sh]: ${MAX_PODS_PER_NODES}

Comment (grayluck) [reply cluster/gce/config-default.sh]: Done.

**Comment by grayluck** [comment]
Thanks for review!

**Pull request merged**